core/src/data_sources/data_source.rs

```diff
@@ -620,8 +698,38 @@ impl DataSource {
             };
 
             for (s, v) in chunk.into_iter().zip(v.into_iter()) {
-                e.push((offset, s, v));
-                offset += 1;
+                let mut hasher = blake3::Hasher::new();
+                hasher.update(s.as_bytes());
+                let text_only_hash = format!("{}", hasher.finalize().to_hex());
+
+                texthash2embedding.insert(text_only_hash, v);
+            }
+        }
+
+        // Ordered results with offset, stirng and vector
+        let embedded_vectors = splits
+            .iter()
+            .enumerate()
+            .map(|(i, s)| {
+                let mut hasher = blake3::Hasher::new();
```

Comment:
I would start by computing the hash for everyone because I think it's the 3rd time you're computing the hash for some chunks here?